vocabularies/UI.xml

```diff
@@ -1837,6 +1837,31 @@ This switches off the validation of user input with respect to decimals.</String
         <Annotation Term="Core.Description" String="The referenced entity set is the preferred starting point for UIs using this service" />
       </Term>
 
+      <Term Name="RecursiveHierarchy" BaseTerm="Hierarchy.RecursiveHierarchy" Type="UI.RecursiveHierarchyType" AppliesTo="EntityType">
+        <Annotation Term="Common.Experimental" />
+        <Annotation Term="Core.Description" String="Rules for hierarchy nodes that a UI can guide the user towards" />
+        <Annotation Term="Core.LongDescription">
+          <String>For example, dropping a node on another node can be prevented if the dropped node cannot
+          be a child of the target node.</String>
+        </Annotation>
+      </Term>
+      <ComplexType Name="RecursiveHierarchyType">
+        <Annotation Term="Common.Experimental" />
+        <Property Name="ChildRules" Type="Collection(UI.RecursiveHierarchyChildType)" Nullable="false">
```

Comment:
Given an entity type that
- has a a NodeType property NT
- and is annotated with UI.RecursiveHierarchy where the Hierarchy.RecursiveHierarchy/NodeTypeProperty points to NT

Imagine NT would take three values A, B and C. How would I define that B and C can be children of A, but does not apply the other way around?

I think a ParentType is also needed, which would lead to: ChildRules is a collection of (ParentType, collection of (RecursiveHierarchyChildType))

Reply:
```json
"ChildRules": [
  {"ChildType": "B", "Allowed": {"$Eq": [{"$Path": "NodeType"}, "A"]}},
  {"ChildType": "C", "Allowed": {"$Eq": [{"$Path": "NodeType"}, "A"]}}
]
```

Reply:
With this proposal above, a single lookup is required to determine the answer for a _given child node_.

We can also look at it from the opposite direction like this:
```
"ChildRules": [
  { "ParentType": "A", "AllowedChildren": ["$In": [   { "$Path": "NodeType"  }, [ "B", "C" ] ] ] } ]
``` 

With this approach, a single lookup is required to determine the answer for a _given parent target node_.

If the use cases at hand have a "preferred" direction, we could select the better fitting approach.  

Maybe a question to @BerSie if there is such a preference.

Reply:
But then it should be `ParentRules`, not `ChildRules`.

Reply:
You're so fast ... I' ve to ask at least one silly question: you are writing "{"$Path": "NodeType"}" - shouldn't it be _NT_ instead of _NodeType_?
Wrt the question of direction - for the drag'n'drop use case we have the child type and with that we know which parents are possible targets (and which not). The same is valid if there is per type a create function. If there is only one create and the child type is selected later on, we have the parent type and the user can choose from the valid child types. The scenario I currently know has the first two use cases.
And there is another question - how to express the extreme cases: all children can be created under one parent, nothing can be created under one parent and all parents can serve as parent for this child type, none of the parents can serve as a parent for this child type (depending on the type of direction we choose).